src/main/groovy/io/github/cosmicsilence/scalafix/ScalafixPlugin.groovy

```diff
@@ -90,31 +87,43 @@ class ScalafixPlugin implements Plugin<Project> {
             }))
             scalafixTask.mode = taskMode
-            scalafixTask.scalaVersion = sourceSet.scalaVersion.get()
-            scalafixTask.classpath = sourceSet.fullClasspath.collect { it.path }
-            scalafixTask.compileOptions = sourceSet.compilerOptions
-            scalafixTask.semanticdbConfigured = extension.isSemanticdbEnabled()
+            scalafixTask.sourceSetName = sourceSet.name
+            scalafixTask.scalaVersion.set(project.provider({ resolveScalaVersion(sourceSet, extension) }))
+            scalafixTask.classpath.set(project.provider( { sourceSet.fullClasspath.collect { it.path } }))
+            scalafixTask.compileOptions.set(project.provider({ sourceSet.compilerOptions }))
+            scalafixTask.semanticdbConfigured = extension.semanticdbEnabled
 
-            if (extension.isSemanticdbEnabled()) {
+            if (extension.semanticdbEnabled) {
                 scalafixTask.dependsOn sourceSet.compileTask
             }
         })
 
         mainTask.dependsOn taskProvider
     }
 
-    private void configureSemanticdbCompilerPlugin(Project project, ScalaSourceSet sourceSet) {
-        def semanticDbCoordinates = ScalafixProps.getSemanticDbArtifactCoordinates(sourceSet.scalaVersion.get(),
-                Optional.ofNullable(project.scalafix.semanticdb.version.getOrNull()))
+    private void configureSemanticdbCompilerPlugin(Project project, ScalaSourceSet sourceSet, ScalafixExtension extension) {
+        def scalaVersion = resolveScalaVersion(sourceSet, extension)
+
+        if (scalaVersion == null || scalaVersion.blank) {
+            throw new GradleException("Unable to detect the Scala version for the '${sourceSet.name}' source set. " +
+                    "Please inform it via the 'scalaVersion' property in the scalafix extension or consider adding " +
+                    "'${sourceSet.name}' to 'ignoreSourceSets'")
+        }
+
+        def semanticDbVersion = Optional.ofNullable(extension.semanticdb.version.getOrNull())
+        def semanticDbCoordinates = ScalafixProps.getSemanticDbArtifactCoordinates(scalaVersion, semanticDbVersion)
         def semanticDbDependency = project.dependencies.create(semanticDbCoordinates)
         def configuration = project.configurations.detachedConfiguration(semanticDbDependency).setTransitive(false)
         def compilerOpts = [
                 '-Xplugin:' + configuration.asPath,
                 '-P:semanticdb:sourceroot:' + project.projectDir,
                 '-Yrangepos'
         ]
-        // intentionally mutating the Scala compile task here to avoid that the Semanticdb compiler plugin
-        // always gets configured and runs even when the Scalafix task is not run (which can be costly).
         sourceSet.addCompilerOptions(compilerOpts)
     }
+
+    private String resolveScalaVersion(ScalaSourceSet sourceSet, ScalafixExtension extension) {
+        def extScalaVersion = extension.scalaVersion.getOrElse("")
```

Comment:
i think using null here and the elvis operator might be easier then relying on blank which is very close to null.

Reply:
The goal was to allow the condition in the ternary operator to rely solely on `.blank`. I'm changing this to check only for null/empty in the ternary operator as `.blank` is not available in older versions of Groovy.

`.getOrElse{ ... }` isn't lazy, but `orElse{ ... }` is. The resulting code would be slightly longer (although negligible), so I think I will stick with the ternary operator for now.

Reply:
I ended up changing this ☝️  a bit after realising I could perform the validation only in one place: this class